Application/ControlMsrE2/ControlMsrE2Hii.c

```diff
@@ -0,0 +1,397 @@
+
+  HiiString = HiiGetString (Context->EfiHandle, Context->IfrOneOf->Question.Header.Prompt, "en-US");
+  if (HiiString) {
+    Print (L"\nBIOS Option found: %s\n", HiiString);
+    FreePool (HiiString);
+  }
+
+  HiiString = AsciiStrCopyToUnicode ((CHAR8 *) Context->IfrVarStore->Name, 0);
+
+  if (HiiString == NULL) {
+    Print (L"\nCould not allocate memory for HiiString\n");
+    return;
+  }
+
+  VarSize = sizeof (EFI_IFR_ONE_OF) - Context->IfrOneOf->Header.Length;
+  VarSize = 8 - (VarSize / 3);
+
+  Print (
+    L"In VarStore \"%s\" GUID: %g Offset: %04X Size: %X ",
+    HiiString,
+    &Context->IfrVarStore->Guid,
+    Context->IfrOneOf->Question.VarStoreInfo.VarOffset,
+    VarSize
+    );
+
+  DataSize = 0;
+
+  Status = gRT->GetVariable (
+                  HiiString,
+                  (VOID *) &Context->IfrVarStore->Guid,
+                  &Attributes,
+                  &DataSize,
+                  NULL
+                  );
+
+  if (Status == EFI_BUFFER_TOO_SMALL) {
+    Data = AllocatePool (DataSize);
+    if (Data != NULL) {
+      Status = gRT->GetVariable (
+                      HiiString,
+                      (VOID *) &Context->IfrVarStore->Guid,
+                      &Attributes,
+                      &DataSize,
+                      Data
+                      );
+
+      if (!EFI_ERROR (Status)) {
+        VarPointer = Data + Context->IfrOneOf->Question.VarStoreInfo.VarOffset;
+        switch (VarSize) {
+        case 1:
+          VarStoreValue = *VarPointer;
+          break;
+        case 2:
+          VarStoreValue = *(UINT16 *) (VarPointer);
+          break;
+        case 4:
+          VarStoreValue = *(UINT32 *) (VarPointer);
+          break;
+        default:
+          VarStoreValue = *(UINT64 *) (VarPointer);
+          break;
+        }
+
+        Print (L"Value: value %X\n", VarStoreValue);
+
+        if (IS_INTERACTIVE ()) {
+          NewValue = VarStoreValue != 0 ? 0 : 1;
+        } else if (IS_LOCK ()) {
+          NewValue = 1;
+        } else if (IS_UNLOCK ()) {
+          NewValue = 0;
+        } else {
+          NewValue = VarStoreValue;
+        }
+
+        if (NewValue != VarStoreValue) {
+          if (IS_INTERACTIVE ()) {
+            Print (L"Do you want to toggle the value y/n ?");
+          } else if (IS_LOCK ()) {
+            Print (L"Do you want to set the value y/n ?");
+          } else {
+            Print (L"Do you want to clear the value y/n ?");
+          }
+
+          if (ReadYN ()) {
+            switch (VarSize) {
+              case 1:
+                *VarPointer = (UINT8)NewValue;
+                break;
+              case 2:
+                *(UINT16 *) (VarPointer) = (UINT16)NewValue;
+                break;
+              case 4:
+                *(UINT32 *) (VarPointer) = (UINT32)NewValue;
+                break;
+              case 8:
+                *(UINT64 *) (VarPointer) = NewValue;
+                break;
+              default:
```

Comment:
Above you assumed that default == 8, please remain consistent either way. Maybe ASSERT.